### ISSUE TITLE
Brush stroke input delay (Windows)

### ISSUE THREAD

**Issue by BrunoDSL** [mypaint/mypaint#803]
l have stylus lagging and delayed stroke response...only on Windows. The Linux version has no such delay. I've been experiencing this problem since 1.1, which forces me to use 1.0 on Windows if I want the stylus to respond regularly, as it should.

MyPaint version: 1.1-1.3
Operating system: Windows 7-10 x86/x64
Wacom Bamboo MTE 450 with the latest drivers.

1. Start MyPaint
2. Try to use your stylus.

**Comment by Playermet** [comment]
Does lag occurs with visible effects (like silver ring)? Try [this](http://www.volnapc.com/all-posts/how-to-get-rid-of-those-annoying-circles-from-your-wacom-cursor-in-windows-7).

**Comment by MR4Y** [comment]
The lag happens regardless of the Windows tablet features being enabled or not.

**Comment by achadwick** [comment]
Is this a duplicate of the earlier lag bugs? #516 + #576 + search? Can you have a dig through and see if it looks like anything similar?

**Comment by achadwick** [comment]
For what it's worth, I never get any lag on my Windows devices, one of which is some ancient dual-core Dell studio 1555, and I never have. I just can't replicate this one, full stop.

Admittedly, I don't *use* Windows on a day to day basis, so I'll be running way less stuff than anyone who does. The only conclusion I personally draw about tablet support on them is to go for the latest Wacom drivers, and that *huge* amounts of suck can be directly attributed to how good the Wintab implementation is for a given driver. The Wacom ones are good and deliver events at ~200 Hz; the Huion ones I've tried are OK but not stellar, and the Genius ones are horrid, slow and broken.

But those differences aren't *lag*, more like a liney-ness of drawing that's down to the brush engine being starved of fine-grained event data.

**Comment by achadwick** [comment]
@MR4Y 
Can you have a go with the latest version and the inking tool please? I want to know if _that's_ laggy when you move control points (you'll see it redraw when you move one of the dots, and it won't be instant, but it should be *fairly* close to real-time unless you are using an especially fancy brush).

:book: *Inking Tool docs:* https://github.com/mypaint/mypaint/wiki/v1.2-Inking-Tool

**Comment by MR4Y** [comment]
I'm already using the latest Wacom drivers. As for the Inking Tool, it takes a few seconds to redraw when you move one control point. Also, the lag happens when you draw with the mouse too, not just the stylus. I'm currently using 1.0, because it's impossible to draw anything on 1.2 due to the lag.

**Comment by MR4Y** [comment]
What other information you need?

**Comment by achadwick** [comment]
I don't know yet whether the lag is of the same severity with the inking tool.

Please can you try the latest beta build? Newer libs may fix bugs.

It's at https://ci.appveyor.com/project/achadwick/mypaint/branch/master (click on either the `MINGW64` or the `MINGW32` job, your call, then click on the "`ARTIFACTS`" tab on the right, and download the `-installer.exe`.

**Comment by MR4Y** [comment]
Well, I tested noth builds and it seems that the lag only happens when the window is centered on the screen. If I drag the window to the side a bit or tile it on either side, the lag is gone.

I then tested with the current release of MyPaint and it has the same behavior.

**Comment by achadwick** [comment]
@MR4Y 
Centred and fullscreened, centred and maximized, or something else? Sometimes desktops do odd rendering pipeline things with fullscreen windows...

Also, is the window size bigger in your "window is centered" state?

**Comment by MR4Y** [comment]
The window is whatever size it sets by default when you first launch the program. There's no difference between maximized and fullscreen windows in terms of behavior. If the window is quite small (I'd say, 25% of the size it would be maximized), it doesn't lag until it starts getting closer to the maximized size (1280x1024 in my case. A bit less height if you discard the taskbar.)

**Comment by achadwick** [comment]
@MR4Y
Can you try another thing please? With the most recent build you have can you try running the included "Event Axes" GTK test program, maximizing the area with the crosshairs and blobs (see below, the one that also shows the tablet name), and then moving the cursor rapidly across it with the pen down?  Do you see the same kind of lag between the crosshair motion and the stylus at all?

(When you're testing, please paste the exact version number from the version you try each time. I know it's tedious, but we have to ask.)

![2017-01-22 gtk3-demo event_axes screenshot](https://user-images.githubusercontent.com/61299/30335427-40eed2f2-97da-11e7-90af-1d5953171b92.png)

**Comment by MR4Y** [comment]
Test results:

- 1.2.1+gitexport.bcf5a28d - Demo is unresponsive and has to be forcefully closed via Task Manager.

- 1.3.0-alpha+git.4e3245de - Pretty much the same behavior as the program. When the window is small, there's virtually no delay. The crosshairs distance from the cursor roughly 3 mm with fast movement. When the window is tiled to either side, fast movement distances the cursor from the crosshairs about 5 mm. When the window is maximized, that distance goes well over 10 mm with fast movement.

**Comment by achadwick** [comment]
(^ additional info: 4e3245de's build is logged at https://ci.appveyor.com/project/achadwick/mypaint/branch/master/job/hkw29x64d1alt13r, and it uses mingw-w64-x86_64-gtk3-3.22.9-1)

**Comment by MR4Y** [comment]
Still not enough info?

**Comment by AtsusaKaneytza** [comment]
I'm having similar issue with my Surface 3 (non-Pro) with MyPaint 1.3.0-alpha+git.0b19fce4 64-bit, but the issue is still persistent in 32-bit on Windows as well. I'm using Windows 10, 64-bit with 2GB of RAM. I used the Event Axes program like you said above and I have no lag when making strokes with full pen pressure. in the Event Axes app. In addition: for me, it doesn't matter how much I reduce the Window size of MyPaint, I experience lag with the brush strokes regardless.
Today, I tested the strokes with a NuVision pen (2048 pressure levels) in addition to my Surface 3 pen (only 256) and the result was the same. Laggy brush strokes. Also sorry if I'm echoing an issue that was already reported.
